### PR TITLE
[MIN-80] Create a notebook for monitoring

### PR DESCRIPTION
**Pull request by osw282**
This PR implements a Jupyter notebook which demonstrate how we can query our database for readability or the embedding drift data store in our postgres metric database using our metric service. Furthermore, the notebook also show a simple example of what we can do with the data we retrieved. A plot for readability over time and a plot for embedding drift over time.

A [README.md](https://github.com/fuzzylabs/MindGPT/pull/59/files#diff-debceb25ab710297f5f192568ce4da9288072641ffb4a8acde38d8bcd00ec995) has also been add which shows how to set up the notebook with our poetry environment. 

NOTE:
New dependency added, run `poetry install`.

### CODE REVIEW

**Review by d-lowl — CHANGES_REQUESTED** [review]
The notebook has some magic values, so it'd be good to add the docs (or point to the existing ones) on how to retrieve the metrics server IP (and probably run it locally?)

In the notebook itself, or in the notebook README

File: notebook/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# 🏃 How do I get started with the monitoring notebook?
```
Comment (JonoCX) [inline notebook/README.md:1]: We probably want a reference to this in the main README

File: notebook/README.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+To start the notebook with our poetry environment, we will first need to install our poetry kernel for Jupyter notebook so that we can access it within the notebook.
+
+
+First, let's find out the name of our environment.
+
+```bash
+poetry env list
```
Comment (JonoCX) [inline notebook/README.md:11]: The presumption here is that there's already active virtual environment?

File: notebook/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+# We should expect to see something like this:
+mindgpt-JiFggUPm-py3.10 (Activated)
+```
+
+Now, we will make our poetry environment accessible from within the notebook by installing it as a ipython kernel.
+
+```bash
+poetry run ipython kernel install --user --name=<KERNEL_NAME>
```
Comment (JonoCX) [inline notebook/README.md:21]: What's the `KERNEL_NAME` here?

Comment (dudeperf3ct) [inline notebook/README.md:21]: Question: Does it not work if we do `poetry shell` and followed by `jupyter notebook`?

File: notebook/monitoring_notebook.ipynb
Comment (dudeperf3ct) [inline notebook/monitoring_notebook.ipynb:1]: Typo, `yplease` to `please` on line 3 of cell 1. Surrounding text, `If you are not sure whether you have set up ...`

File: pyproject.toml [hunk 4]
```diff
@@ -37,6 +37,7 @@ pytest = "^7.2.2"
 pytest-cov = "^4.0.0"
 mypy = "1.1.1"
 typos = "^1.16.1"
+matplotlib = "^3.7.2"
```
Comment (dudeperf3ct) [inline pyproject.toml:40]: I would consider this in a dev group dependency, the difference being, the `main` dependencies required for running the project and the `dev` dependencies required for the development of this project.

File: notebook/README.md (hunk 3 above)
Comment (osw282) [reply notebook/README.md:21]: I don't think so. I tried `jupyter notebook` in poetry shell, the env didn't come up.

File: notebook/monitoring_notebook.ipynb
Comment (osw282) [reply notebook/monitoring_notebook.ipynb:1]: Not sure where this is? In the notebook, not in README?

File: pyproject.toml (hunk 4 above)
Comment (osw282) [reply pyproject.toml:40]: Good idea.

Comment (osw282) [reply pyproject.toml:40]: Actually, I think this is already in the dev group dependency.

File: notebook/monitoring_notebook.ipynb
Comment (dudeperf3ct) [reply notebook/monitoring_notebook.ipynb:1]: In the notebook, reviewing notebook is tough, so I have added surrounding text as hint where the typo is located as well.

**Review by dudeperf3ct — APPROVED** [review]
LGTM

**Review by JonoCX — CHANGES_REQUESTED** [review]
Nice work - see comments, mostly focused on the README

**Review by d-lowl — APPROVED** [review]

File: notebook/README.md (hunk 2 above)
Comment (osw282) [reply notebook/README.md:11]: Added a command to activate the poetry shell first.

File: README.md [hunk 5]
```diff
@@ -226,6 +226,8 @@ To compute the embedding drift when running the embedding pipeline, we will port
 kubectl port-forward service/monitoring-service 5000:5000
 ```
 
+There is also a [notebook](notebook/monitoring_notebook.ipynb) that demonstrate how we could retrieve our metric data with the metric service and a simple example of what we can do with the data.
```
Comment (JonoCX) [inline README.md]: I'd make this stand out a bit more, e.g:

### Monitoring MindGPT 👀 

We've created a notebook which accesses the monitoring service, fetches the metrics, and creates some simple plots showing the change over time. 

This is a starting point for accessing the metrics, and we're planning to introduce a hosted dashboard version of these plots at some point in the future.

**Review by JonoCX — CHANGES_REQUESTED** [review]
Nice - just a small change to the main readme to highlight the presence of the monitoring notebook

**Pull request merged**